pkg/util/limiter/query_limiter.go

```diff
@@ -0,0 +1,89 @@
+package limiter
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/pkg/labels"
+	"go.uber.org/atomic"
+
+	"github.com/cortexproject/cortex/pkg/cortexpb"
+	"github.com/cortexproject/cortex/pkg/ingester/client"
+	"github.com/cortexproject/cortex/pkg/util"
+	"github.com/cortexproject/cortex/pkg/util/validation"
+)
+
+type QueryLimiter struct {
+	uniqueSeriesMx sync.RWMutex
+	uniqueSeries   map[model.Fingerprint]struct{}
+
+	chunkBytesCount   *atomic.Int32
+	maxSeriesPerQuery int
+}
+type queryLimiterCtxKey struct{}
+
+var (
+	qlCtxKey        = &queryLimiterCtxKey{}
+	errMaxSeriesHit = "The query hit the max number of series limit while fetching chunks %s (limit: %d)"
+)
```

Comment:
[nit] Move this above `type QueryLimiter struct` so we keep the struct definition close to its `NewQueryLimiter()` initialiser.